ethereumj-core/src/main/java/org/ethereum/sync/SyncManager.java

```diff
@@ -155,6 +165,184 @@ public void init(final ChannelManager channelManager, final SyncPool pool) {
         }
     }
 
+    @Override
+    protected void blockRetrieveLoop() {
+        class BlocksCallback implements FutureCallback<List<Block>> {
+
+            private Channel peer;
+
+            public BlocksCallback(Channel peer) {
+                this.peer = peer;
+            }
+
+            @Override
+            public void onSuccess(List<Block> result) {
+                addBlocks(result, peer.getNodeId());
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                logger.debug("{}: Error receiving Blocks. Dropping the peer.", name, t);
+                peer.getEthHandler().dropConnection();
+            }
+        }
+
+        List<SyncQueueIfc.BlocksRequest> bReqs = emptyList();
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                if (bReqs.isEmpty()) {
+                    bReqs = syncQueue.requestBlocks(16 * 1024).split(MAX_IN_REQUEST);
+                }
+
+                if (bReqs.isEmpty() && headersDownloadComplete) {
+                    logger.info("{}: Block download complete.", name);
+                    finishDownload();
+                    downloadComplete = true;
+                    return;
+                }
+
+                int blocksToAsk = getBlockQueueFreeSize();
+                if (blocksToAsk >= MAX_IN_REQUEST) {
+//                    SyncQueueIfc.BlocksRequest bReq = syncQueue.requestBlocks(maxBlocks);
+
+                    if (bReqs.size() == 1 && bReqs.get(0).getBlockHeaders().size() <= 3) {
+                        // new blocks are better to request from the header senders first
+                        // to get more chances to receive block body promptly
+                        for (BlockHeaderWrapper blockHeaderWrapper : bReqs.get(0).getBlockHeaders()) {
+                            Channel channel = pool.getByNodeId(blockHeaderWrapper.getNodeId());
+                            if (channel != null) {
+                                ListenableFuture<List<Block>> futureBlocks
+                                        = channel.getEthHandler().sendGetBlockBodies(singletonList(blockHeaderWrapper));
+
+                                if (futureBlocks != null) {
+                                    // print needed to prevent an infinite loop when there are 2 blocks for download
+                                    logger.trace("futureBlocks number: " + futureBlocks.get().size());
```

Comment:
Also we are trying to use `logger.trace` only with `if (logger.isTraceEnabled()`